## 1.Below is the table containing the information of Customer price index, discounts, offers where the organization wants to predict the sales based on the cpi, discounts,offers.

| CPI      | Discounts |  Offers  |  Sales   | 
|---------:|:---------:|---------:|---------:|
|  2600    |    3      |   20     |  550000  |  
|  3000    |    4      |   15     |  565000  | 
|  3200    |    5      |   18     |  610000  | 
|  3600    |    3      |   30     |  595000  |  
|  4000    |    5      |    8     |  760000  | 
|  4100    |    6      |    8     |  810000  | 

In [6]:
import pandas as pd
import numpy as np
import plotly.express as px
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore|")

# Linear Regression 

In [3]:
data = pd.read_excel('Data set.xlsx', sheet_name=0)
data

CPI  discounts  offers   Sales
0  2600          3      20  550000
1  3000          4      15  565000
2  3200          5      18  610000
3  3600          3      30  595000
4  4000          5       8  760000
5  4100          6       8  810000

In [8]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression()

In [9]:
X = data.iloc[:, :-1]
y = data.iloc[:, -1]

In [10]:
X

CPI  discounts  offers
0  2600          3      20
1  3000          4      15
2  3200          5      18
3  3600          3      30
4  4000          5       8
5  4100          6       8

In [11]:
y

0    550000
1    565000
2    610000
3    595000
4    760000
5    810000
Name: Sales, dtype: int64

In [12]:
reg.fit(X, y)

LinearRegression()

In [15]:
y_pred = reg.predict(X)
y_pred

array([518401.04171196, 600201.34257367, 617094.24971522, 597810.70928584,
       768867.81214403, 787624.84456927])

In [18]:
from sklearn.metrics import r2_score
r2_score(y, y_pred)

0.9517053277318955

In [19]:
reg.score(X, y)

0.9517053277318955

### Given below information find out the Sales that has
### 1. 5000 cpi , 3 percentage discounts, 20 rewards offers  


In [23]:
y_pred_1 = reg.predict([[5000, 3, 20]])
y_pred_1

array([826645.34838222])

### 2. 4000 cpi , 8 percentage discounts, 19 rewards offers

In [24]:
y_pred_2 = reg.predict([[4000, 8, 19]])
y_pred_2

array([732680.36486005])

### We have used the Linear Regression model It has given 95.17% accuracy
### 1. 5000 cpi , 3 percentage discounts, 20 rewards offers - Sales is 826645.34838222
### 2. 4000 cpi , 8 percentage discounts, 19 rewards offers - Sales is 732680.36486005
### This is very Good Model

# KNeighbor Regressor

In [29]:
from sklearn.neighbors import KNeighborsRegressor
kreg = KNeighborsRegressor()

In [30]:
kreg.fit(X, y)

KNeighborsRegressor()

In [31]:
y_pred = kreg.predict(X)
y_pred

array([616000., 616000., 616000., 668000., 668000., 668000.])

In [32]:
r2_score(y, y_pred)

0.3121500559910415

### Given below information find out the Sales that has
### 1. 5000 cpi , 3 percentage discounts, 20 rewards offers  

In [33]:
y_pred_1 = kreg.predict([[5000, 3, 20]])
y_pred_1

array([668000.])

### 2. 4000 cpi , 8 percentage discounts, 19 rewards offers

In [34]:
y_pred_2 = kreg.predict([[4000, 8, 19]])
y_pred_2

array([668000.])

### We have used the KNeighbor Regressor model It has given 31.21% accuracy
### 1. 5000 cpi , 3 percentage discounts, 20 rewards offers - Sales is 668000
### 2. 4000 cpi , 8 percentage discounts, 19 rewards offers - Sales is 668000
### This is  Bad Model

# Random Forest Regressor

In [35]:
from sklearn.ensemble import RandomForestRegressor
rreg = RandomForestRegressor()

In [36]:
rreg.fit(X, y)

RandomForestRegressor()

In [38]:
y_pred = rreg.predict(X)
y_pred

array([562000., 576250., 610300., 596500., 738650., 766100.])

In [39]:
r2_score(y, y_pred)

0.9553874300111982

### Given below information find out the Sales that has
### 1. 5000 cpi , 3 percentage discounts, 20 rewards offers  

In [40]:
y_pred_1 = rreg.predict([[5000, 3, 20]])
y_pred_1

array([649400.])

### 2. 4000 cpi , 8 percentage discounts, 19 rewards offers

In [41]:
y_pred_2 = rreg.predict([[4000, 8, 19]])
y_pred_2

array([680450.])

### We have used the Random Forest Regressor model It has given 95.53% accuracy
### 1. 5000 cpi , 3 percentage discounts, 20 rewards offers - Sales is 649400
### 2. 4000 cpi , 8 percentage discounts, 19 rewards offers - Sales is 680450
### This is Very Good Model

# Conclusion

## We have performed 3 Regression Model 
## 1. Linear Regression (Multiple Linear Regression) with 95.17% accuracy(r2 score)
## 2. KNearest neighbor Regressor with 31.21% accuracy(r2 score)
## 3. Random Forest Regressor with 95.53% accuracy(r2 score)

### Out of these 3, ***KNeighbour Regressor*** performed very poorly so we can't choose that model, but with other model we can go with either ***Linear Regression*** model or ***Random Forest Regressor*** both are good.